### DISCUSSION TITLE
3D images?

### DISCUSSION THREAD

**Discussion by caiuspetronius** [plenoptic-org/plenoptic#354] [Ideas]
It looks like only 2D images are implemented.  Was there any work done toward implementing 3D steerable pyramids, etc? E.g., along the lines of http://cdelleluche.chez-alice.fr/SPIEarticle.pdf. It would be of interest in areas like medical imaging, seismology, material science, etc.

**Comment by billbrod** [comment]
Thanks for your question! We have not worked on a 3d steerable pyramid implementation as part of plenoptic. This is interesting, and @pehf has said he's interested in, but no one has the bandwidth to look into it. We are not currently planning on implementing it, but if someone wants to take a crack at it, we'd be willing to review it and consider it for inclusion. Do you know of existing implementations elsewhere?

**Reply by caiuspetronius** [reply]
No, I didn't find anything either in Matlab or Python.

**Reply by billbrod** [reply]
Huh, well that makes it more difficult. We'll update here if we start working on it, but like I said, it's not currently planned.

**Comment by caiuspetronius** [comment]
Hi William,

I attached my take on 3D steerable pyramids. Equations in the paper that I
cited seem to be missing some parts, so I implemented something similar but
not identical fixing the normalization problems, etc. Seems to work well,
so take a look if you are interested.

Best,
Yury

**Reply by billbrod** [reply]
Hi Yury, no implementation is attached. It sounds like maybe you responded to the github via your email? I think github strips out attachments when you do that. If you write your comment on github directly, you can either attach a file (or paste the contents).

**Reply by caiuspetronius** [reply]
Oh, sorry. Here is the link, https://github.com/caiuspetronius/steerablePyramid3D/blob/main/steerablePyramid3D.py

**Reply by billbrod** [reply]
Got it, thanks for sharing. This will be a helpful starting point if we decide to work on this for plenoptic.

**Reply by caiuspetronius** [reply]
Sure, my pleasure. I'm now working on implementing the Simoncelli's textureAnalyse and textureSynthesize functions for 3D images, this is what I'm after actually. I'll update the github repo once these are finished and tested.

**Reply by billbrod** [reply]
Good luck! If you haven't checked out the texture model currently in plenoptic (https://github.com/plenoptic-org/plenoptic/blob/main/src/plenoptic/simulate/models/portilla_simoncelli.py), I'd use that as a reference (rather than the original matlab code), because (in my biased opinion) it's easier to understand. Happy to answer any questions about that implementation.

If you write the 3d Texture model in pytorch (which would also require writing the 3d pyramid in pytorch), you can just use plenoptic's metamer object to synthesize the metamers, rather than writing your own textureSynthesize function.

**Reply by caiuspetronius** [reply]
Thanks, I'll take a look. Their matlab code is a bit dense, but I've figured it out by now.

**Reply by caiuspetronius** [reply]
Hi William, yes, your code looks pretty straightforward. I already wrote the texture analysis function, but it is easy to make it into the same model class as yours. Are you saying that your metamers synthesis would just synthesise a 3D image given a 3D model?

**Reply by billbrod** [reply]
It should, yes! We haven't tested 3D images much in plenoptic, and the optimization is harder (because there are more pixels), but everything should work*. If it doesn't, I'd appreciate you letting me know! We want to support this type of use, but haven't gotten around to fully testing it.

*:  something that definitely won't work are the various helper functions for displaying metamers -- they're all meant to work with 2d images, rather than 3d images. We have some support for videos, but that won't work for you.

**Reply by caiuspetronius** [reply]
Ok, I'll give it a try then. Do I understand correctly that you are using an entirely different optimization route from the Simoncelli's? Something along the lines of treating each input pixel as a model parameter and using the standard Torch optimization route?

**Reply by caiuspetronius** [reply]
Because I was thinking of training a generator NN, like a VNet, to take a random image and generate an output that has the right set of statistics metrics to match a given texture. This way one could reuse the net to get more textures much faster.

**Reply by billbrod** [reply]
Yes, that is correct -- we treat the pixels as parameters and use standard `torch.optim` classes (by default, Adam, but you can change this) to update the pixels so that the model outputs are the same. We don't get the exact same texture synthesis outputs as the original algorithm, but we believe they're of comparable perceptual quality. You can see [some examples](https://docs.plenoptic.org/docs/branch/main/tutorials/models/Metamer-Portilla-Simoncelli.html) in our docs and compare to the original paper.

**Reply by caiuspetronius** [reply]
I see. It would be interesting to compare. Because the Simoncelli's optimization has convergence issues when using a large number of angles.

**Reply by billbrod** [reply]
Interesting -- let me know what you find!

**Reply by caiuspetronius** [reply]
Hi William, I implemented https://github.com/caiuspetronius/steerablePyramid3D/blob/main/steerablePyramid3D.py in torch. So, now  it takes tensors with optional dimensions in front of the image to allow parallel training on batches and multiple pixel channels. Also, the pyramid expansion and texture statistics function can be used for 3D/2D/1D data now.

**Reply by billbrod** [reply]
Great! In order to work with plenoptic, the model should be an object that inherits `torch.nn.Module`, with an `__init__` and a `forward` method (you can see more detail in [our docs](https://docs.plenoptic.org/docs/branch/main/models.html)). That will require a little bit of reworking of your code -- I think largely just splitting up your `__main__` into those two methods. You can see plenoptic's texture implementation to see how I do that. Happy to answer any questions you have or to point out some more resources if you haven't written torch modules before.

**Reply by caiuspetronius** [reply]
Hi William, I wonder what you think about using the Simoncelli's asymmetric (one-lobe) oriented filters instead of their symmetric counterparts as in "Simplified Design of Steerable Pyramid Filters" K. R. Castleman, M. Schulze, Q. Wu. I implemented both (there are also small differences in their radial functions), and I find that Simoncelli's are hard to justify. For example, in the example image that I decomposed below there is a saw-tooth pattern. Because it is purely asymmetric one would expect the real parts of all the oriented components nearly zero with all the power in the imaginary components. And this is indeed the case with the symmetric Castleman filters. With Simoncelli's filters, on the other hand, there is about equal power in real and imaginary parts. You can see this in the attached plots of the amplitudes of the components. The Castelman's amplitudes have lots of sharp features coming from the imaginary parts, while the Simoncelli's amplitudes are all fuzzy because the real and imaginary components are shifted wrt each other by half the period and blur out when added in the amplitude plots. I compared my decomposition results with plenoptic ones and they look identical except that mine are not downsampled (I choose to keep full-res versions and do the downsampling only for the autocorrelation computations).

Castelman's filters:
<img width="912" alt="Screenshot 2025-06-11 at 9 40 00 AM" src="https://github.com/user-attachments/assets/8888effc-6997-4881-be2c-c5e6b8d66448" />

Simoncelli's filters:
<img width="912" alt="Screenshot 2025-06-11 at 9 39 20 AM" src="https://github.com/user-attachments/assets/f5f6a81b-aa12-4036-bafc-ee83dd0a8fe3" />

**Reply by caiuspetronius** [reply]
Because there are lots of statistics calculated based on the magnitude of the components it looks like lots of spatial information is lost by the Simoncelli's filters.

**Comment by billbrod** [comment]
Can you include the plenoptic code you ran, and the code you used to generate the sawtooth image? I'm not sure what the last row in your figures is showing; when I reconstruct a sinusoid using only one of the orientations (which I thought is what you're showing), I don't get anything like that.

For asymmetric vs. symmetric: the real and imaginary components of the pyramid will be symmetric / anti-symmetric (cosine / sine). Which is which depends on the order of the filters / number of orientations in the pyramid: [see docstring](https://github.com/plenoptic-org/plenoptic/blob/50f810388af6998e17914c30714b610cf695e4bf/src/plenoptic/simulate/canonical_computations/steerable_pyramid_freq.py#L66).

(note that the filters are symmetric/anti-symmetric a.k.a. [even/odd](https://en.wikipedia.org/wiki/Even_and_odd_functions); none of the filters are asymmetric)

It sounds like you're computing what we (and the Portilla-Simoncelli paper) call magnitudes, the complex modulus (`sqrt(real^2 + imag^2)`). This intentionally throws away that local spatial information by averaging across the sine and cosine phases of the filters. In the texture model, when we need that precise spatial information, we just use the real components (this is used to compute the phase alignment across scales, for example).

So whether you want the magnitude, which has thrown away that precise info, or just one of the phases, which has not, depends on your use case, and both are used in the texture model.

(and as a note, you can get the non-downsampled pyramid from plenoptic by passing `downsample=False` when initializing it, though that's not what we use in the texture model)

**Comment by caiuspetronius** [comment]
Hi William, I rewrote my code to subsample the lower-spatial scale components like in the original PS paper. This made it at least 2x faster in calculation of the statistics vector, so it was worth it: https://github.com/caiuspetronius/steerablePyramid3D/tree/main. The code to generate the sawtooth image is in Matlab:

imsize = 128;
v = [ 1, 1, 1 ];
P = 8;

[ X, Y, Z ] = meshgrid( 1 : imsize, 1:imsize, 1:imsize );
R = sqrt( ( X - imsize / 2 ).^2 + ( Y - imsize / 2 ).^2 + ( Z - imsize / 2 ).^2 );

I = mod( X * v( 1 ) + Y * v( 2 ) + Z * v( 3 ), P );
I = I / max( I, [], 'all' );
ave = mean( I, 'all' );
I = ( I - ave ) * 0.25 + ave;
tmp = 0.25 * ( rand( size( I ) ) - ave ) + ave;
innr = imsize / 2 - 10;
outr = imsize / 2 - 5;
I( R > innr ) = tmp( R > innr );
I( R > outr ) = 0;

figure, imshow( I( :, :, imsize / 2 ), [ 0.25, 0.75 ] )

nii = make_nii( I, [ 1, 1, 1 ] );
save_nii( nii, 'sawtooth3D_img.nii.gz' );

I included the generated 3D nifty image into the repo so you can read it like in the __main__  function of steerablePyramid.py and save the central slice that I was using into something like png for your testing. Or you can just use the sawtooth.pgm in Simoncelli's example images to the same end.

I am not computing magnitudes in the steerablePyramid.py, I was just plotting the mags. Here are all plots of the components (real, imag, and mag) for the PS filters and just the mag plot for the CS filters. The rows in the figures show orientation components (4) for each scale except for the first column which I used to plot the original, LP, HP, and the reconstructed images. The images were generated in the __main__ function of the steerablePyramid.py, so you can get all the details from there.

PS real:
![sawtooth_2D_real_Simoncelli](https://github.com/user-attachments/assets/92b76e9a-3a39-4020-b79b-23cda54aacb9)

PS imag:
![sawtooth_2D_imag_Simoncelli](https://github.com/user-attachments/assets/d6a77886-d400-4041-9bb5-ef2f9cc0d579)

PS mag:
![sawtooth_2D_abs_Simoncelli](https://github.com/user-attachments/assets/8b7a4295-0e1f-48ad-98f6-c372380a8023)

CS mag:
![sawtooth_2D_abs](https://github.com/user-attachments/assets/5d9132b2-37a4-4a96-95b7-c4a4cc21f892)

You can see that CS (both pos and neg lobes of the cosine orientation filter) mags have more detail than the PS (only the pos lobes) mags. This is because imag and real parts of the PS expansion are pi/2 shifted wrt each other (see images). When added in quadrature they combine into blurry mags. In the CS expansion the real parts of the components are zero for the sawtooth image and so they don't blur the imag parts. 

Another thing: the PS radial functions are not suitable for 1D data while the CS radial functions are (see the example of a 1D signal reconstructions in the  repo's README.md. It doesn't reconstruct correctly with PS radial functions because they are non-smooth).

**Reply by billbrod** [reply]
For the CS filters, if the real and imag parts are not phase-shifted with respect to each other, how are they related?

Like I said before, for us, the fact that you get the "blurry mags" is part of the point -- you are discarding the precise spatial information you get when using a single phase. This is similar to the relationship between simple and complex cells in the primate primary visual cortex: simple cells respond to a single phase (like the real component alone), whereas complex cells are phase-invariant (like the magnitude), and so are invariant to small changes in position.

Good to know about the 1D case: we've largely used these filters on 2D images, so we haven't tested them on other cases.

**Comment by caiuspetronius** [comment]
Also, I noticed that all the textures generated from random noise (by my code or by yours or by Simoncelli's Matlab code) have residual high spatial frequency noise left and it is quite noticeable. This is apparently because the autocorrelation of the HP component is not in the statistics vector, only its variance. I wondered why, so I added the autocorrelation as well and it did make the noise smaller (and the final loss smaller too). The residual noise can also be reduced by increasing the weights of all 3 autocorrelation statistics (autocorr_mag, autocorr_recon, autocorr_HP) in the statistics vector by a factor of 10 for example.

**Reply by billbrod** [reply]
For this issue, in [our documentation](https://docs.plenoptic.org/docs/branch/main/tutorials/models/Metamer-Portilla-Simoncelli.html), we initialize synthesis using `im_init = (torch.rand_like(img) - 0.5) * 0.1 + img.mean()` (where `img` is the texture to match) -- that is, uniform noise that goes from mean +/- 0.05. (I've increased that to +/- 0.1 with decent success as well), which I think avoids this issue (I do not remember if this is what the original paper does). The changes you suggest also seem like a reasonable way to handle this.
